ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/HD/LMDB.hs

```diff
@@ -0,0 +1,499 @@
+      LIInitialiseFromMemory slot vals -> (GDDMustNotExist, pure (), initFromVals slot vals)
+      _ -> (GDDMustExist, pure (), \_ -> pure ())
+
+  -- get the filepath for this db creates the directory if appropriate
+  dbFilePath <- guardDbDir gdd sfs path
+
+  -- copy from another lmdb path if appropriate
+  copy_db_action
+
+  -- open this database
+  dbEnv <- liftIO $ LMDB.openEnvironment dbFilePath limits
+
+  -- the LMDB.Database that holds the DbState (i.e. sequence number)
+  -- This transaction must be read-write because on initialisation it creates the database
+  dbSettings <- liftIO $ LMDB.readWriteTransaction dbEnv $ LMDB.getDatabase (Just "_dbstate")
+
+  -- TODO: at some point Javier was able to get an LMDB which didn't have this
+  -- database. How is it possible? maybe some copy function is not copying this
+  -- one?
+
+  -- Here we get the LMDB.Databases for the tables of the ledger state
+  -- Must be read-write transaction because tables may need to be created
+  dbBackingTables <- liftIO $ LMDB.readWriteTransaction dbEnv $
+    traverseLedgerTables getDb namesLedgerTables
+
+  let
+    db = Db{..}
+    bsClose :: m ()
+    bsClose = do
+      Trace.traceWith dbTracer $ TDBClosing dbFilePath
+      open_handles <- IOLike.atomically $ IOLike.readTVar dbOpenHandles
+      for_ open_handles vhClose
+      liftIO $ LMDB.closeEnvironment dbEnv
+      Trace.traceWith dbTracer $ TDBClosed dbFilePath
+
+    bsCopy shfs (HD.BackingStorePath to0) = do
+      to <- guardDbDir GDDMustNotExist shfs to0
+      lmdbCopy dbTracer dbEnv to
+
+    bsValueHandle = mkLMDBBackingStoreValueHandle db
+
+    bsWrite :: SlotNo -> LedgerTables l DiffMK -> m ()
+    bsWrite slot diffs = do
+      old_slot <- liftIO $ LMDB.readWriteTransaction dbEnv $ withDbSettingsRW dbSettings $ \s@DbState{dbsSeq} -> do
+        -- TODO This should be <. However the test harness does call bsWrite with the same slot
+        unless (dbsSeq <= At slot) $ Exn.throw $ DbErrNonMonotonicSeq (At slot) dbsSeq
+        void $ zipLedgerTablesA writeLMDBTable dbBackingTables diffs
+        pure (dbsSeq, s {dbsSeq = At slot})
+      Trace.traceWith dbTracer $ TDBWrite old_slot slot
+
+  -- now initialise those tables if appropriate
+  init_action db
+
+  Trace.traceWith dbTracer $ TDBOpened dbFilePath
+  pure HD.BackingStore{..}
+
+
+data TraceValueHandle
+  = TVHOpened
+  | TVHClosing
+  | TVHClosed
+  | TVHSubmissionStarted
+  | TVHSubmissionEnded
+  deriving stock(Show, Eq)
+
+mkValueHandle :: forall m. (MonadIO m, IOLike m)
+  => Trace.Tracer m TraceDb
+  -> LMDB.Environment LMDB.ReadOnly
+  -> IOLike.TVar m (Map Int (ValueHandle m))
+  -> m (ValueHandle m)
+mkValueHandle dbTracer0 dbEnv dbOpenHandles = do
+  let LMDB.Env env = dbEnv
+  readOnlyTx <- liftIO $ LMDB.mdb_txn_begin env Nothing (LMDB.isReadOnlyEnvironment dbEnv)
+
+  (r, traces) <- IOLike.atomically $ IOLike.stateTVar dbOpenHandles $ \x0 ->
+    let
+      vh_id = maybe 0 ((+1) . fst) $ Map.lookupMax x0
```

Comment:
Matter of taste: Extract this conversion to value handle ID to a top-level function?